src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+                        // foreach(...)
+                        // return list;
+                        var symbolName1 = GetFreeSymbolNameAndMarkUsed(variableName);
+                        variable = SyntaxFactory.IdentifierName(symbolName1);
+                        nodesBefore = new[] { CreateLocalDeclarationStatement(symbolName1, initializer) };
+                        nodesAfter = new[] { SyntaxFactory.ReturnStatement(variable).WithAdditionalAnnotations(Simplifier.Annotation) };
+                        return true;
+                    case SyntaxKind.Argument:
+                        // before return SomeMethod((from a in b select a).ToList());
+                        // after var list = new List<T>();
+                        // foreach(...)
+                        // SomeMethod(list);
+                        var symbolName2 = GetFreeSymbolNameAndMarkUsed(variableName);
+                        variable = SyntaxFactory.IdentifierName(symbolName2);
+                        nodesBefore = new[] { CreateLocalDeclarationStatement(symbolName2, initializer) };
+                        nodesAfter = new[] { parentStatement.ReplaceNode(invocationExpression, variable.WithAdditionalAnnotations(Simplifier.Annotation)) };
+                        return true;
+                        // SyntaxKind.ArrowExpressionClause is not supported
+                }
+
+                nodesBefore = default;
+                nodesAfter = default;
+                variable = default;
+                return false;
+            }
+
+            private LocalDeclarationStatementSyntax CreateLocalDeclarationStatement(SyntaxToken identifier, ExpressionSyntax initializer)
+            {
+                return SyntaxFactory.LocalDeclarationStatement(
+                            SyntaxFactory.VariableDeclaration(
+                                VarNameIdentifier,
+                                SyntaxFactory.SingletonSeparatedList(
+                                    SyntaxFactory.VariableDeclarator(
+                                        identifier,
+                                        argumentList: null,
+                                        SyntaxFactory.EqualsValueClause(initializer))))).WithAdditionalAnnotations(Simplifier.Annotation);
+            }
+
+            private bool TryReplaceWithLocalFunction(out DocumentUpdateInfo documentUpdateInfo)
+            {
+                var parentStatement = _source.GetAncestorOrThis<StatementSyntax>();
+                if (parentStatement != null)
+                {
+                    StatementSyntax internalNodeMethod(ExpressionSyntax expression)
+                        => SyntaxFactory.YieldStatement(SyntaxKind.YieldReturnStatement, expression);
+
+                    if (TryConvertQueryExpression(internalNodeMethod, out var convertedFromQueryExpression))
+                    {
+                        if (TryGetTypeSyntax(_source, out var typeSyntax))
```

Comment:
>typeSyntax [](start = 62, length = 10)

Do we assume this type is ```IEnumerable<T>```? I don't think it has to be, but it mast be ```IEnumerable<T>``` on the local function.

Reply:
Probably we should get converted type from the expression in the last Select and simply use it as ```T``` for ```IEnumerable<T>```

---
In reply to: [178406693](https://github.com/dotnet/roslyn/pull/25362#discussion_r178406693) [](ancestors = 178406693)

Reply:
> 1. TryGetTypeSyntax exactly gets T from queryExpression. I believe it is the same as the latest select.

To me it looks like TryGetTypeSyntax simply gets the type returned by the query. It doesn't have to be IEnumerable.
For example from Queryable extensions:
```
public static IQueryable<TResult> Select<TSource, TResult>(
 this IQueryable<TSource> source,
 Expression<Func<TSource, TResult>> selector
)
```
To me, it looks like this refactoring will never do the right thing for a query that targets Queryable. 

In general, the type doesn't even have to be generic. 

For the local function transformation we need to figure out what is ```T``` for ```IEnumerable<T>```. Because we need return type of local function to be ```IEnumerable<T>``` regardless of what type query returns. I suggest to infer that ```T``` from sector of the last ```select```. Similar how you do that for ```foreach``` local. I.e. by grabbing ConvertedType from the select expression.


>. 2. We cannot use ConvertedType. See  ReturnObject. It returns object there.

Local function doesn't return object in that test. This is an unrelated type.


---
In reply to: [178422801](https://github.com/dotnet/roslyn/pull/25362#discussion_r178422801) [](ancestors = 178422801,178407054,178406693)